### PR TITLE
DHCP client fixes

### PR DESCRIPTION
**Pull request by troglobit**
This PR address the problems reported in issues #278 and #283.

### CODE REVIEW

File: src/confd/src/infix-dhcp.c [hunk 1]
```diff
@@ -22,6 +22,25 @@ static const struct srx_module_requirement reqs[] = {
...
+        i++;
+
+    while (i < len && !isspace(arg[i]) && j < len - 1)
+        buf[j++] = buf[i++];
+
+    buf[j] = '\0';
+
+    return buf;
+}
+
```
Comment (wkz) [inline src/confd/src/infix-dhcp.c:43]: Can't we enfore this with a `must` expression?

File: board/common/rootfs/usr/share/udhcpc/default.script [hunk 2]
```diff
@@ -105,7 +105,13 @@ case "$ACTION" in
 	    /usr/sbin/avahi-autoipd -c $interface && /usr/sbin/avahi-autoipd -k $interface
 	fi
 
-	if /bin/ip addr add dev $interface $ip/$NETMASK $BROADCAST proto 5; then
+	if [ -n "$NETMASK" ]; then
+	    IP=$ip/$NETMASK
+	else
+	    IP=$ip
```
Comment (wkz) [inline board/common/rootfs/usr/share/udhcpc/default.script:111]:
```suggestion
	    IP=$ip/32
```
I had to try this manually (address without prefix) to see what iproute2 would do. Suggest we make it clear to future readers what will happen in this case.

File: board/common/rootfs/usr/share/udhcpc/default.script [hunk 3]
```diff
@@ -105,7 +105,13 @@ case "$ACTION" in
...
+	fi
+
+	if /bin/ip addr add dev $interface $IP $BROADCAST proto 5; then
```
Comment (wkz) [inline board/common/rootfs/usr/share/udhcpc/default.script:114]: Do we have a ticket to fix symbolic protocol name lookups?

File: board/common/rootfs/usr/share/udhcpc/default.script [hunk 4]
```diff
@@ -121,6 +121,17 @@ case "$ACTION" in
...
 	set_dhcp_routes
 
+	# set hostname if given
+	if [ -n "$hostname" ]; then
+	    log "setting new hostname: $hostname"
+	    hostname "$hostname"
+
+	    if grep -q "^127.0.1.1.*" /etc/hosts; then
+		log "Updating /etc/hosts with new hostname"
+		sed -i "s/^\(127.0.1.1\s\+\).*/\1$hostname/" /etc/hosts
```
Comment (wkz) [inline board/common/rootfs/usr/share/udhcpc/default.script:131]: Any utility in using `aug` here? (Like we do in `confd`)

File: src/confd/src/infix-dhcp.c (hunk 1 above)
Comment (troglobit) [reply src/confd/src/infix-dhcp.c:43]: This is for handling option value exceptions, like hostname, other options can have free form text string values, or hex.  It's definitely doable, but it'll take a while to figure out.

File: board/common/rootfs/usr/share/udhcpc/default.script (hunk 2 above)
Comment (troglobit) [reply board/common/rootfs/usr/share/udhcpc/default.script]: Good point, committed your suggestion, but will probably set the default NETMASK earlier in a follow-up.

File: board/common/rootfs/usr/share/udhcpc/default.script (hunk 3 above)
Comment (troglobit) [reply board/common/rootfs/usr/share/udhcpc/default.script:114]: You mean the `proto 5` ugliness? No, not that I'm aware of.

File: board/common/rootfs/usr/share/udhcpc/default.script (hunk 4 above)
Comment (troglobit) [reply board/common/rootfs/usr/share/udhcpc/default.script:131]: Heh, forgot about aug. I'll have a look, that would be useful also for one of our customer cases as well.

File: src/confd/src/infix-dhcp.c (hunk 1 above)
Comment (wkz) [reply src/confd/src/infix-dhcp.c:43]: I got curios :angel:. This seems to work:

```yang
	  must "../name != 'hostname' or re-match(., '[a-zA-Z0-9\\-_]{0,253}')";
```

Comment (troglobit) [reply src/confd/src/infix-dhcp.c:43]: Ha ha ha, OK thank you! I'll push another update :bowing_man::pray:

File: board/common/rootfs/usr/share/udhcpc/default.script (hunk 4 above)
Comment (troglobit) [reply board/common/rootfs/usr/share/udhcpc/default.script]: Easy peasy with augtool :sunglasses:

File: board/common/rootfs/usr/share/udhcpc/default.script (hunk 3 above)
Comment (troglobit) [reply board/common/rootfs/usr/share/udhcpc/default.script]: New issue #285 tracked by new project.

**Review by wkz — APPROVED** [review]

**Pull request merged**